common/resource.go

```diff
@@ -108,9 +122,38 @@ func DecodeSignalMsg(raw []byte) (string, interface{}, error) {
 			err := json.Unmarshal([]byte(msg.Payload), &answer)
 			return msg.ReplyTo, answer, err
 		case SignalMsgICE:
-			var candidates []webrtc.ICECandidate
+			var candidates []ICECandidate
+			var webRTCCandidates []webrtc.ICECandidate
 			err := json.Unmarshal([]byte(msg.Payload), &candidates)
-			return msg.ReplyTo, candidates, err
+			if err != nil {
+				return msg.ReplyTo, webRTCCandidates, err
+			}
+
+			for _, c := range candidates {
```

Comment:
Not sure how this even worked before, but the payload contains the ICE Typ as a string, yet here it is expected to be an int. This function here is the culprit
```go
func (t ICECandidateType) String() string {
	switch t {
	case ICECandidateTypeHost:
		return iceCandidateTypeHostStr
	case ICECandidateTypeSrflx:
		return iceCandidateTypeSrflxStr
	case ICECandidateTypePrflx:
		return iceCandidateTypePrflxStr
	case ICECandidateTypeRelay:
		return iceCandidateTypeRelayStr
	default:
		return ErrUnknownType.Error()
	}
}
```

Reply:
I'm a little skeptical of this change.  I'm worried it will actually break the relationship between producers (uncensored users / widgets) and consumers (censored users / Lantern clients).

Here's why:
- Looking at [Freddie](https://github.com/getlantern/browsersunbounded/blob/ecef5f622fef12af1d92b004030f8ba0a1ed0447/freddie/freddie.go), it seems that ICE signaling messages originate on consumers/producers and are simply forwarded by Freddie.
- Your changes here only affect the decoding of signaling messages, not the encoding.
- The version of `pion/webtrc` used by this repo hasn't changed in over a year.  [Here](https://github.com/getlantern/browsersunbounded/blob/1566f60bd3f51217ff8994004537734fbfab4631/go.mod#L12) the version specified one year ago and [here](https://github.com/getlantern/browsersunbounded/blob/88d46b1bfb8ed8768f98cbcf519b98911fa7da24/go.mod#L12) is the version specified on your branch.
- We know that widgets and clients are currently able to peer and establish a tunnel to the egress server (the one running on Heroku).

Given all of this, why would deploying an egress server on Lantern Cloud break the encoding format of messages sent by widgets?  The egress server should have no effect on this IIUC.

Would you mind explaining more about how you arrived at this change?  Could you provide a test or something which demonstrates how the code is currently broken?  Could you also explain why peers are currently able to connect without your changes?

Reply:
I can post the error I saw that prompted me to add this change. I'll post that tomorrow

Reply:
@hwh33 this is the error. 
```
Error decoding signal message: json: cannot unmarshal string into Go struct field ICECandidate.type of type webrtc.ICECandidateType (msg: {"ReplyTo":"d0186037-a7ae-4f3a-acc7-0d6adbdd3812","Type":3,"Payload":"[{\"foundation\":\"1952023002\",\"priority\":2130706431,\"address\":\"192.168.1.64\",\"protocol\":1,\"port\":50471,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"364963319\",\"priority\":2130706431,\"address\":\"2c0f:3d00:615:7d00:1841:d569:e079:eae2\",\"protocol\":1,\"port\":59101,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1896410339\",\"priority\":2130706431,\"address\":\"2c0f:3d00:615:7d00:8eb:cdf7:9182:b498\",\"protocol\":1,\"port\":52262,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1413034161\",\"priority\":2130706431,\"address\":\"100.102.26.53\",\"protocol\":1,\"port\":64079,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"3533976542\",\"priority\":2130706431,\"address\":\"fd7a:115c:a1e0::7d26:1a35\",\"protocol\":1,\"port\":53796,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"2589410501\",\"priority\":2130706431,\"address\":\"192.168.64.1\",\"protocol\":1,\"port\":52169,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1219771392\",\"priority\":2130706431,\"address\":\"fd0c:c9cb:8354:42db:1cbc:f197:1d2e:e9da\",\"protocol\":1,\"port\":61749,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1241098944\",\"priority\":1694498815,\"address\":\"196.250.65.87\",\"protocol\":1,\"port\":65080,\"type\":\"srflx\",\"component\":1,\"relatedAddress\":\"0.0.0.0\",\"relatedPort\":60155,\"tcpType\":\"\"}]"}
)
```

Reply:
@hwh33 at this point I am not sure if this error is not occurring in production. I can quite easily reproduce this error using the `github.com/pion/webrtc/v3 v3.2.6` which is the version used currently. The code below reproduces
 
go.mod
-------
```
module github/AGMETEOR/test_go

go 1.21.4

require github.com/pion/webrtc/v3 v3.2.6

require (
	github.com/davecgh/go-spew v1.1.1 // indirect
	github.com/google/uuid v1.3.1 // indirect
	github.com/pion/datachannel v1.5.8 // indirect
	github.com/pion/dtls/v2 v2.2.12 // indirect
	github.com/pion/ice/v2 v2.3.36 // indirect
	github.com/pion/interceptor v0.1.29 // indirect
	github.com/pion/logging v0.2.2 // indirect
	github.com/pion/mdns v0.0.12 // indirect
	github.com/pion/randutil v0.1.0 // indirect
	github.com/pion/rtcp v1.2.14 // indirect
	github.com/pion/rtp v1.8.7 // indirect
	github.com/pion/sctp v1.8.19 // indirect
	github.com/pion/sdp/v3 v3.0.9 // indirect
	github.com/pion/srtp/v2 v2.0.20 // indirect
	github.com/pion/stun v0.6.1 // indirect
	github.com/pion/transport/v2 v2.2.10 // indirect
	github.com/pion/turn/v2 v2.1.6 // indirect
	github.com/pmezard/go-difflib v1.0.0 // indirect
	github.com/stretchr/testify v1.9.0 // indirect
	github.com/wlynxg/anet v0.0.3 // indirect
	golang.org/x/crypto v0.21.0 // indirect
	golang.org/x/net v0.22.0 // indirect
	golang.org/x/sys v0.18.0 // indirect
	gopkg.in/yaml.v3 v3.0.1 // indirect
)

```

main.go
--------
```
package main

import (
	"encoding/json"
	"fmt"

	"github.com/pion/webrtc/v3"
)

const (
	SignalMsgGenesis SignalMsgType = iota
	SignalMsgOffer
	SignalMsgAnswer
	SignalMsgICE
)

type SignalMsgType int

func (t SignalMsgType) String() string {
	switch t {
	case SignalMsgGenesis:
		return "Genesis"
	case SignalMsgOffer:
		return "Offer"
	case SignalMsgAnswer:
		return "Answer"
	case SignalMsgICE:
		return "ICE"
	default:
		return "invalid"
	}
}

type SignalMsg struct {
	ReplyTo string
	Type    SignalMsgType
	Payload string
}

func DecodeSignalMsg(raw []byte) (string, interface{}, error) {
	var err error
	var msg SignalMsg

	err = json.Unmarshal(raw, &msg)

	if err == nil {
		switch msg.Type {
		case SignalMsgAnswer:
			var answer webrtc.SessionDescription
			err := json.Unmarshal([]byte(msg.Payload), &answer)
			return msg.ReplyTo, answer, err
		case SignalMsgICE:
			var webRTCCandidates []webrtc.ICECandidate
			err := json.Unmarshal([]byte(msg.Payload), &webRTCCandidates)
			if err != nil {
				return msg.ReplyTo, webRTCCandidates, err
			}

			return msg.ReplyTo, webRTCCandidates, err
		}
	}

	return "", nil, err
}

func main() {
	iceBytes := []byte(`{"ReplyTo":"5733058b-c875-4d40-a012-fd585eb48605","Type":3,"Payload":"[{\"foundation\":\"1952023002\",\"priority\":2130706431,\"address\":\"192.168.1.64\",\"protocol\":1,\"port\":59707,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"364963319\",\"priority\":2130706431,\"address\":\"2c0f:3d00:615:7d00:1841:d569:e079:eae2\",\"protocol\":1,\"port\":52832,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1896410339\",\"priority\":2130706431,\"address\":\"2c0f:3d00:615:7d00:8eb:cdf7:9182:b498\",\"protocol\":1,\"port\":59400,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1413034161\",\"priority\":2130706431,\"address\":\"100.102.26.53\",\"protocol\":1,\"port\":49751,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"3533976542\",\"priority\":2130706431,\"address\":\"fd7a:115c:a1e0::7d26:1a35\",\"protocol\":1,\"port\":49809,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"2589410501\",\"priority\":2130706431,\"address\":\"192.168.64.1\",\"protocol\":1,\"port\":57447,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1219771392\",\"priority\":2130706431,\"address\":\"fd0c:c9cb:8354:42db:1cbc:f197:1d2e:e9da\",\"protocol\":1,\"port\":51055,\"type\":\"host\",\"component\":1,\"relatedAddress\":\"\",\"relatedPort\":0,\"tcpType\":\"\"},{\"foundation\":\"1241098944\",\"priority\":1694498815,\"address\":\"196.250.65.87\",\"protocol\":1,\"port\":64786,\"type\":\"srflx\",\"component\":1,\"relatedAddress\":\"0.0.0.0\",\"relatedPort\":63091,\"tcpType\":\"\"}]"}`)

	_, _, err := DecodeSignalMsg(iceBytes)
	if err != nil {
		fmt.Println("Error decoding signal message:", err)
		return
	}
}
```

Bumping the pion/webrtc version to `3.3.4` fixes the issue